### ISSUE TITLE
Datasources: Introduce a max size for datasource responses

### ISSUE THREAD

**Issue by bergquist** [grafana/grafana#35796]
Some backend datasources consume quite a lot of memory when processing the datasources responses. This results in unpredictable memory usages for Grafana instances which is undesirable when operating Grafana at scale. Considering the data returned from the datasource should be representable within a dashboard I can't see any reason why we should allow the Grafana server to process responses larger than X bytes. 

I suggest we introduce a new setting for the dataproxy/datasource that returns an error if the response is too big. I would suggest implementing a new middleware using https://golang.org/pkg/io/#LimitReader in the datasource cache and make the max size configurable.

**Comment by marefr** [comment]
Could it be added as middleware to the HTTP client provider ([code](https://github.com/grafana/grafana/blob/16850630bf766213e824b93eb8119d8932864c3f/pkg/infra/httpclient/httpclientprovider/http_client_provider.go#L21)) so that it would apply for both data proxy and backend HTTP requests?

Is there really a problem querying a lot of data through the data proxy, it should be streamed to the response I believe? For backend data sources like Elasticsearch I can see the problem since it reads the whole response into memory before parsing it.

**Comment by bergquist** [comment]
> Is there really a problem querying a lot of data through the data proxy,

Not really. The main problem is backend datasources as they consume memory. But I think enforcing it in both data proxy and data sources using one setting might make sense.

**Comment by marefr** [comment]
SQL data sources is a big problem so this is not only related to HTTP data sources.

**Comment by marefr** [comment]
Outgoing HTTP requests (prio 1):
I think it should be possible to implement a middleware to limit responses and register that with the [HTTP Client Provider](https://github.com/grafana/grafana/blob/main/pkg/infra/httpclient/httpclientprovider/http_client_provider.go).

Was looking at https://golang.org/pkg/io/#LimitReader and that seems to work, however as I understand it will return `io.EOF` when limit is reached. Might not be optimal since this would not signal that something went wrong. In the example of a data source expecting JSON response and getting limited with `io.EOF` it would still try and parse the response as JSON which might yield some other weird errors.

As an alternative I looked at https://github.com/nanmu42/limitio which seems to allow returning a specific error in the case of limit reached. This seems to me like a better solution. Not sure we should use this library though since it's very small, but we can perhaps use the `ReadCloser` as inspiration?

SQL data sources (prio 1):
I think we need a setting to control the maximum rows allowed, rather than number of bytes allowed - at least I don't see a way controlling amount of bytes read with SQL. Today we have a constant controlling the maximum amount of rows returned which is huge - think replacing this with a setting would allow us to limit things:
https://github.com/grafana/grafana/blob/56f61001a8269a50832b9416a6096e162963e631/pkg/tsdb/sqleng/sql_engine.go#L144 

External backend data sources (prio 2):
It's a bit tricky to enforce things there, but we can at least try to do it for our own plugins. For example we've heard there's been some memory problems requesting to much data using the Google Sheets data source plugin. If we implement a middleware for outgoing HTTP requests it should probably live in the SDK to allow "any" HTTP data source to use it. This implies that the response limit setting needs to be propagated to external backend plugins. Before actually making use of this we probably need a better story in the SDK of how you actually use the HTTP client provider in an external backend plugin. We could basically decide to move things to the SDK whenever we want, from the beginning or later.

**Comment by dsotirakis** [comment]
>Might not be optimal since this would not signal that something went wrong. In the example of a data source expecting JSON response and getting limited with io.EOF it would still try and parse the response as JSON which might yield some other weird errors.

@marefr, what if we can check the initial response size and then compare it to the `io.EOF` signal we get, and decide if we should throw an error or if it's actually EOF?

**Comment by dsotirakis** [comment]
Or - there's also https://pkg.go.dev/net/http#MaxBytesReader which is under `net/http` we can use. 

According to the docs,
> MaxBytesReader is similar to io.LimitReader but is intended for limiting the size of incoming request bodies. In contrast to io.LimitReader, MaxBytesReader's result is a ReadCloser, returns a non-EOF error for a Read beyond the limit, and closes the underlying reader when its Close method is called.
MaxBytesReader prevents clients from accidentally or maliciously sending a large request and wasting server resources.

**Comment by marefr** [comment]
> Or - there's also https://pkg.go.dev/net/http#MaxBytesReader which is under net/http we can use.

I was under the impression that's only suitable for limiting responses of incoming requests to the server, but I might be wrong. Just give it a go, try things out and see what works or not.

**Comment by marefr** [comment]
I added an issue for SQL data souces #38975

**Comment by bergquist** [comment]
> In the example of a data source expecting JSON response and getting limited with io.EOF it would still try and parse the response as JSON which might yield some other weird errors.

Another option would be to create a new reader containing the error message and override the it in the response if the limit is reached.

**Comment by marefr** [comment]
@bergquist that's what we've basically already done in #38962

**Comment by marefr** [comment]
Few more things left before we can close this.

**Comment by marefr** [comment]
After consideration I think it's better to open a new issue (epic) tracking all the sub tasks.
